### PR TITLE
[PT FE]: support aten::instance_norm

### PR DESCRIPTION
**Pull request by eaidova**
### Details:
 - *iaten::instance_norm*
 - *...*

### Tickets:
 - *101051*

### CODE REVIEW

File: src/frontends/pytorch/src/op/instance_norm.cpp [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+            bias = reshape_conv_bias(context, bias, norm);
+            norm = context.mark_node(std::make_shared<ov::op::v1::Add>(norm, bias));
+        }
+        return {norm};
+    }
+    auto batch_dim = context.mark_node(std::make_shared<ov::op::v1::Gather>(input_shape, zero, zero));
+    auto channel_dim = context.mark_node(std::make_shared<ov::op::v1::Gather>(input_shape, one, zero));
+    auto batch_dim_1d = context.mark_node(std::make_shared<ov::op::v0::Unsqueeze>(batch_dim, zero));
+    auto batch_norm_channels = context.mark_node(std::make_shared<ov::op::v1::Multiply>(batch_dim, channel_dim));
+    auto batch_norm_channels_1d = context.mark_node(std::make_shared<ov::op::v0::Unsqueeze>(batch_norm_channels, zero));
```
Comment (mvafin) [inline src/frontends/pytorch/src/op/instance_norm.cpp:55]:
```suggestion
    auto batch_norm_channels_1d = context.mark_node(std::make_shared<ov::op::v1::Multiply>(batch_dim_1d, channel_dim));
```
Result should be 1d with auto broadcast

**Review by mvafin — CHANGES_REQUESTED** [review]

File: src/frontends/pytorch/src/op/instance_norm.cpp [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+#include "openvino/op/tile.hpp"
+#include "openvino/op/unsqueeze.hpp"
+#include "utils.hpp"
+
+namespace ov {
+namespace frontend {
+namespace pytorch {
+namespace op {
+
+OutputVector translate_instance_norm(NodeContext& context) {
```
[the remark this answers is not in the document]
Comment (eaidova) [reply src/frontends/pytorch/src/op/instance_norm.cpp:26]: added

File: src/frontends/pytorch/src/utils.cpp [hunk 3]
```diff
@@ -13,6 +13,16 @@ namespace ov {
 namespace frontend {
 namespace pytorch {
 
+void num_inputs_check(NodeContext& context, int min_inputs, int max_inputs) {
+    auto inputs = context.inputs();
+    FRONT_END_OP_CONVERSION_CHECK(inputs.size() < min_inputs, "Got less inputs than expected");
```
Comment (eaidova) [inline src/frontends/pytorch/src/utils.cpp:18]:
```suggestion
    FRONT_END_OP_CONVERSION_CHECK(inputs.size() > min_inputs, "Got less inputs than expected");
```

File: src/frontends/pytorch/src/op/instance_norm.cpp [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+#include "openvino/op/convert_like.hpp"
+#include "openvino/op/gather.hpp"
+#include "openvino/op/multiply.hpp"
+#include "openvino/op/mvn.hpp"
+#include "openvino/op/range.hpp"
+#include "openvino/op/reshape.hpp"
+#include "openvino/op/shape_of.hpp"
+#include "openvino/op/squeeze.hpp"
+#include "openvino/op/tile.hpp"
+#include "openvino/op/unsqueeze.hpp"
```
Comment (rkazants) [inline src/frontends/pytorch/src/op/instance_norm.cpp:18]: @mvafin, include the single header will be much better in this case:)

File: src/frontends/pytorch/src/op/instance_norm.cpp [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+namespace frontend {
+namespace pytorch {
+namespace op {
+
+OutputVector translate_instance_norm(NodeContext& context) {
+    num_inputs_check(context, 8, 9);
+    auto input = context.get_input(0);
+    auto eps = static_cast<float>(context.const_input<double>(7));
+    auto input_shape = context.mark_node(std::make_shared<ov::op::v3::ShapeOf>(input));
+    auto rank_1d = context.mark_node(std::make_shared<ov::op::v3::ShapeOf>(input_shape));
```
Comment (rkazants) [inline src/frontends/pytorch/src/op/instance_norm.cpp:31]: can you please align opset version for all nodes in the decomposition and use the latest opset?

File: src/frontends/pytorch/src/op/instance_norm.cpp [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+#include "openvino/op/multiply.hpp"
+#include "openvino/op/mvn.hpp"
+#include "openvino/op/range.hpp"
+#include "openvino/op/reshape.hpp"
+#include "openvino/op/shape_of.hpp"
+#include "openvino/op/squeeze.hpp"
+#include "openvino/op/tile.hpp"
+#include "openvino/op/unsqueeze.hpp"
+#include "utils.hpp"
+
```
Comment (rkazants) [inline src/frontends/pytorch/src/op/instance_norm.cpp:20]: please use `using namespace ov::opset10;` that will save from extra `ov::op::`, etc.

File: src/frontends/pytorch/src/utils.cpp [hunk 7]
```diff
@@ -13,6 +13,16 @@ namespace ov {
 namespace frontend {
 namespace pytorch {
 
+void num_inputs_check(NodeContext& context, int min_inputs, int max_inputs) {
```
Comment (rkazants) [inline src/frontends/pytorch/src/utils.cpp:16]:
```suggestion
void num_inputs_check(const NodeContext& context, size_t min_inputs, size_t max_inputs) {
```

File: src/frontends/pytorch/src/utils.cpp [hunk 8]
```diff
@@ -13,6 +13,16 @@ namespace ov {
...
+    auto inputs = context.inputs();
+    FRONT_END_OP_CONVERSION_CHECK(inputs.size() > min_inputs, "Got less inputs than expected");
```
Comment (rkazants) [inline src/frontends/pytorch/src/utils.cpp:18]: type inconsistensy: `int` vs. `size_t`

File: src/frontends/pytorch/src/utils.cpp [hunk 9]
```diff
@@ -13,6 +13,16 @@ namespace ov {
...
+    if (inputs.size() > max_inputs) {
+        for (int i = max_inputs; i < inputs.size(); i++) {
```
Comment (rkazants) [inline src/frontends/pytorch/src/utils.cpp:20]:
```suggestion
        for (size_t i = max_inputs; i < inputs.size(); i++) {
```

File: src/frontends/pytorch/src/utils.cpp (hunk 7 above)
Comment (rkazants) [reply src/frontends/pytorch/src/utils.cpp:16]: Additionally, we can check type of the operation in this routine. This check will prevent us from wrong translator call for some operation.

File: src/frontends/pytorch/src/op/instance_norm.cpp [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+        bias = context.get_input(2);
+        bias = context.mark_node(std::make_shared<ov::op::v0::Tile>(bias, batch_dim_1d));
+    }
+    auto running_mean = context.get_input(3);
+    running_mean = context.mark_node(std::make_shared<ov::op::v0::Tile>(running_mean, batch_dim_1d));
+    auto running_var = context.get_input(4);
+    running_var = context.mark_node(std::make_shared<ov::op::v0::Tile>(running_var, batch_dim_1d));
+    auto batch_norm = context.mark_node(
+        std::make_shared<ov::op::v5::BatchNormInference>(reshaped_input, weight, bias, running_mean, running_var, eps));
+    return {context.mark_node(std::make_shared<ov::op::v1::Reshape>(batch_norm, input_shape, true))};
```
Comment (rkazants) [inline src/frontends/pytorch/src/op/instance_norm.cpp:83]: super mega decomposition. Is there any chance to simply it? That is because `InstanceNormalization` will be represented with large graph:)

File: src/frontends/pytorch/src/utils.cpp (hunk 9 above)
Comment (mvafin) [reply src/frontends/pytorch/src/utils.cpp:20]: I would rather do `auto` in this case

File: src/frontends/pytorch/src/op/instance_norm.cpp (hunk 4 above)
Comment (mvafin) [reply src/frontends/pytorch/src/op/instance_norm.cpp:18]: That is a rare case, for such cases we can agree to use opset or operations on developer decision, but I think this can stay as it is.

File: src/frontends/pytorch/src/op/instance_norm.cpp (hunk 10 above)
Comment (eaidova) [reply src/frontends/pytorch/src/op/instance_norm.cpp:83]: if you have some suggestions, I'm ready to hear.
In majority cases decomposition with BatchNorm used, when instance norm in training mode, what is rare case in our situation, so I think it is not critiical

File: src/frontends/pytorch/src/utils.cpp (hunk 9 above)
Comment (eaidova) [reply src/frontends/pytorch/src/utils.cpp]: fixed

File: src/frontends/pytorch/src/utils.cpp (hunk 8 above)
Comment (eaidova) [reply src/frontends/pytorch/src/utils.cpp:18]: fixed

File: src/frontends/pytorch/src/op/instance_norm.cpp (hunk 10 above)
Comment (rkazants) [reply src/frontends/pytorch/src/op/instance_norm.cpp:83]: let us split these two cases `training` and `inference` into separate helper functions and call one of them based on number inputs.

File: src/frontends/pytorch/src/op/instance_norm.cpp (hunk 6 above)
Comment (eaidova) [reply src/frontends/pytorch/src/op/instance_norm.cpp:21]: applied

File: src/frontends/pytorch/src/op/instance_norm.cpp (hunk 10 above)
Comment (eaidova) [reply src/frontends/pytorch/src/op/instance_norm.cpp]: applied

**Review by rkazants — APPROVED** [review]

**Comment by eaidova** [comment]
@mvafin is there anything else what need to be done?

File: src/frontends/pytorch/src/op/instance_norm.cpp [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+#include "utils.hpp"
+
+namespace ov {
+namespace frontend {
+namespace pytorch {
+namespace op {
+
+using namespace ov::opset10;
+
+OutputVector translate_instance_norm_inference(NodeContext& context,
```
Comment (mvafin) [inline src/frontends/pytorch/src/op/instance_norm.cpp:29]: Please add helper function to nameless namespace

File: src/frontends/pytorch/src/op/instance_norm.cpp [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+    }
+    if (!context.input_is_none(2)) {
+        auto bias = context.get_input(2);
+        bias = reshape_conv_bias(context, bias, norm);
+        norm = context.mark_node(std::make_shared<Add>(norm, bias));
+    }
+    return {norm};
+}
+
+OutputVector translate_instance_norm_train(NodeContext& context,
```
Comment (mvafin) [inline src/frontends/pytorch/src/op/instance_norm.cpp:48]:
```suggestion
OutputVector translate_instance_norm_train(const NodeContext& context,
```

Comment (mvafin) [reply src/frontends/pytorch/src/op/instance_norm.cpp:48]: Please also add to nameless namespace

File: src/frontends/pytorch/src/op/instance_norm.cpp (hunk 11 above)
Comment (mvafin) [inline src/frontends/pytorch/src/op/instance_norm.cpp:29]:
```suggestion
OutputVector translate_instance_norm_inference(const NodeContext& context,
```
Maybe it would be also required to change `reshape_conv_bias` function

File: src/frontends/pytorch/src/utils.cpp [hunk 13]
```diff
@@ -13,6 +13,16 @@ namespace ov {
 namespace frontend {
 namespace pytorch {
 
+void num_inputs_check(NodeContext& context, size_t min_inputs, size_t max_inputs) {
```
Comment (mvafin) [inline src/frontends/pytorch/src/utils.cpp:16]:
```suggestion
void num_inputs_check(const NodeContext& context, size_t min_inputs, size_t max_inputs) {
```

File: src/frontends/pytorch/src/utils.hpp [hunk 14]
```diff
@@ -18,6 +18,8 @@ class FrameworkNode;
 namespace frontend {
 namespace pytorch {
 
+void num_inputs_check(NodeContext& context, size_t min_inputs, size_t max_inputs);
```
Comment (mvafin) [inline src/frontends/pytorch/src/utils.hpp:21]:
```suggestion
void num_inputs_check(const NodeContext& context, size_t min_inputs, size_t max_inputs);
```

File: src/frontends/pytorch/src/utils.cpp [hunk 15]
```diff
@@ -13,6 +13,16 @@ namespace ov {
...
+    auto inputs = context.inputs();
+    FRONT_END_OP_CONVERSION_CHECK(inputs.size() > min_inputs, "Got less inputs than expected");
+    if (inputs.size() > max_inputs) {
```
Comment (mvafin) [inline src/frontends/pytorch/src/utils.cpp:19]: This `if` is not needed here

File: src/frontends/pytorch/src/op/instance_norm.cpp (hunk 11 above)
Comment (eaidova) [reply src/frontends/pytorch/src/op/instance_norm.cpp]: moved

File: src/frontends/pytorch/src/op/instance_norm.cpp (hunk 12 above)
Comment (eaidova) [reply src/frontends/pytorch/src/op/instance_norm.cpp:47]: updated

File: src/frontends/pytorch/src/op/instance_norm.cpp (hunk 11 above)
Comment (eaidova) [reply src/frontends/pytorch/src/op/instance_norm.cpp]: renamed

File: src/frontends/pytorch/src/utils.cpp (hunk 13 above)
Comment (eaidova) [reply src/frontends/pytorch/src/utils.cpp]: updated

File: src/frontends/pytorch/src/utils.hpp (hunk 14 above)
Comment (eaidova) [reply src/frontends/pytorch/src/utils.hpp]: updated

File: src/frontends/pytorch/src/utils.cpp (hunk 15 above)
Comment (eaidova) [reply src/frontends/pytorch/src/utils.cpp]: removed

**Review by mvafin — APPROVED** [review]

**Comment by mvafin** [comment]
@eaidova you have legit build issue https://dev.azure.com/openvinoci/dldt/_build/results?buildId=480319&view=logs&j=44bc7f8f-e82e-568c-0ef2-3321258cabd9&t=f039189d-08dd-5241-4c59-f7fadebdac3e&l=7314
Master merge won't solve this

File: src/frontends/pytorch/src/op/instance_norm.cpp [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+namespace {
+OutputVector translate_instance_norm_inference(const NodeContext& context,
+                                               const Output<Node>& input,
+                                               const Output<Node>& reduction_axes,
+                                               float eps) {
+    auto norm = context.mark_node(
+        std::make_shared<ov::op::v6::MVN>(input, reduction_axes, true, eps, ov::op::MVNEpsMode::INSIDE_SQRT));
+    if (!context.input_is_none(1)) {
+        auto weight = context.get_input(1);
+        weight = reshape_chanelwise(context, weight, norm);
```
Comment (mvafin) [inline src/frontends/pytorch/src/op/instance_norm.cpp:36]:
```suggestion
        weight = reshape_channelwise(context, weight, norm);
```

File: src/frontends/pytorch/src/op/instance_norm.cpp [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+        norm = context.mark_node(std::make_shared<ov::op::v1::Multiply>(norm, weight));
+    }
+    if (!context.input_is_none(2)) {
+        auto bias = context.get_input(2);
+        bias = reshape_chanelwise(context, bias, norm);
```
Comment (mvafin) [inline src/frontends/pytorch/src/op/instance_norm.cpp:41]:
```suggestion
        bias = reshape_channelwise(context, bias, norm);
```

**Review by mvafin — APPROVED** [review]
